Lib/test/test_tabnanny.py

```diff
@@ -0,0 +1,344 @@
+        with TemporaryPyFile(SOURCE_CODES["nannynag_errored"]) as file_path:
+            out = f"{file_path} 3 '\\tprint(\"world\")\\n'\n"
+            self.verify_tabnanny_check(file_path, out=out)
+
+    def test_when_no_file(self):
+        """A python file which does not exist actually in system."""
+        path = 'no_file.py'
+        err = f"{path!r}: I/O Error: [Errno 2] No such file or directory: {path!r}\n"
+        self.verify_tabnanny_check(path, err=err)
+
+    def test_errored_directory(self):
+        """Directory containing wrongly indented python source code files."""
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            error_file = TemporaryPyFile(
+                SOURCE_CODES["wrong_indented"], directory=tmp_dir
+            )
+            code_file = TemporaryPyFile(
+                SOURCE_CODES["error_free"], directory=tmp_dir
+            )
+            with error_file as e_file, code_file as c_file:
+                err = ('unindent does not match any outer indentation level'
+                            ' (<tokenize>, line 3)\n')
+                err = f"{e_file!r}: Indentation Error: {err}"
+                self.verify_tabnanny_check(tmp_dir, err=err)
+
+
+class TestProcessTokens(TestCase):
+    """Testing `tabnanny.process_tokens()`."""
+
+    @mock.patch('tabnanny.NannyNag')
+    def test_with_correct_code(self, MockNannyNag):
+        """A python source code without any whitespace related problems."""
+
+        with TemporaryPyFile(SOURCE_CODES["error_free"]) as file_path:
+            with open(file_path) as f:
+                tabnanny.process_tokens(tokenize.generate_tokens(f.readline))
+            self.assertFalse(MockNannyNag.called)
+
+    def test_with_errored_codes_samples(self):
+        """A python source code with whitespace related sampled problems."""
+
+        # "tab_space_errored_1": executes block under type == tokenize.INDENT
+        #                        at `tabnanny.process_tokens()`.
+        # "tab space_errored_2": executes block under
+        #                        `check_equal and type not in JUNK` condition at
+        #                        `tabnanny.process_tokens()`.
+
+        for key in ["tab_space_errored_1", "tab_space_errored_2"]:
+            with self.subTest(key=key):
+                with TemporaryPyFile(SOURCE_CODES[key]) as file_path:
+                    with open(file_path) as f:
+                        tokens = tokenize.generate_tokens(f.readline)
+                        with self.assertRaises(tabnanny.NannyNag):
+                            tabnanny.process_tokens(tokens)
+
+
+class TestCommandLine(TestCase):
+    """Tests command line interface of `tabnanny`."""
+
+    def validate_cmd(self, *args, stdout="", stderr=""):
+        """Common function to assert the behaviour of command line interface."""
+        _, out, err = script_helper.assert_python_ok('-m', 'tabnanny', *args)
```

Comment:
You may use:

proc = script_helper.spawn_python('-m', 'tabnanny', *args, text=True)
out, err = proc.communicate()
self.assertEqual(out, stdout)
self.assertEqual(err, stderr)
self.assertEqual(proc.returncode, 0)

text=True normalizes newlines to \n (Unix EOL).